### ISSUE TITLE
Erreur

### ISSUE THREAD

**Issue by ogandril** [theislab/cellrank#179]
I have installed the latest version:
Successfully installed cellrank-1.1.dev580+g99ebbf1

And then 
>>> import cellrank as cr
>>> scv.tl.velocity_graph(adata, mode_neighbors='connectivities') 

works fine. And then 

>>> cr.tl.find_root(adata)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'cellrank.tools' has no attribute 'find_root'

Thank's for your help.

Olivier

**Comment by michalk8** [comment]
Hi @ogandril ,
`find_root` has been recently renamed to `root_states`.

**Comment by ogandril** [comment]
The following code:

_import scvelo as scv
import cellrank as cr
adata = scv.read('adata_dynamical.h5ad')
scv.tl.velocity_graph(adata, mode_neighbors='connectivities')
cr.tl.root_states(adata)_

generates the following error:

>>> cr.tl.root_states(adata)
_WARNING: Number of states (17) was automatically determined by `eigengap` but no Schur decomposition was found. Computing with default parameters
Unable to import PETSc or SLEPc.
You can install it from: https://slepc4py.readthedocs.io/en/stable/install.html
Defaulting to `method='brandts'`.
Traceback (most recent call last):
  File "/opt/miniconda3/lib/python3.7/site-packages/cellrank/tools/_root_final.py", line 122, in _root_final
    n_states=n_states, cluster_key=cluster_key, **kwargs
  File "/opt/miniconda3/lib/python3.7/site-packages/cellrank/tools/estimators/_gppca.py", line 477, in compute_metastable_states
    raise e
  File "/opt/miniconda3/lib/python3.7/site-packages/cellrank/tools/estimators/_gppca.py", line 474, in compute_metastable_states
    self._gpcca = self._gpcca.optimize(m=n_states)
  File "/opt/miniconda3/lib/python3.7/site-packages/msmtools/analysis/dense/gpcca.py", line 1323, in optimize
    self._do_schur_helper(max(m_list))
  File "/opt/miniconda3/lib/python3.7/site-packages/msmtools/analysis/dense/gpcca.py", line 1121, in _do_schur_helper
    raise ValueError(f'Clustering into {m} clusters will split conjugate eigenvalues. '
ValueError: Clustering into 17 clusters will split conjugate eigenvalues. Request one cluster more or less. 

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/miniconda3/lib/python3.7/site-packages/cellrank/tools/_root_final.py", line 180, in root_states
    **kwargs,
  File "/opt/miniconda3/lib/python3.7/site-packages/cellrank/tools/_root_final.py", line 126, in _root_final
    f"Computing {n_states} metastable states cuts through a block of complex conjugates. "
TypeError: unsupported operand type(s) for +: 'NoneType' and 'int'_


Since it mention PETSc, I tried to install it:

pip install petsc petsc4py
Installing collected packages: petsc, petsc4py
    Running setup.py install for petsc ... done
    Running setup.py install for petsc4py ... done

But I got the same error running cr.tl.root_states(adata)

**Comment by michalk8** [comment]
@ogandril thanks for spotting this, it wasn't a SLEPc issue, but faulty logic in the code. Let me know if #180 fixed it for you.

**Comment by ogandril** [comment]
OK, the cr.tl.root_states() function now works. 

The cr.tl.find_final() does not; 

>>> cr.tl.find_final(adata)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: module 'cellrank.tools' has no attribute 'find_final'

I guess it also is a naming issue of the function?

**Comment by Marius1311** [comment]
yes! The function is called `cr.tl.final_states` now. Sorry about this, will update the tutorials accordingly.

**Comment by ogandril** [comment]
I had to install slepc and slepc4py, but it now works!! 
Thank's for your help!!!

The result is somewhat mesmerizing though...
Final states more or less in phase with latent time, less so for root states. 
But my interpretation may be incorrect?

![Figure_root_final](https://user-images.githubusercontent.com/18306816/84043071-247c1b80-a9a6-11ea-98ee-9cecf6ee6eba.png)





(base) OG-MBP15T-OG:OG3289 olivier_2$ pip install petsc petsc4py
Requirement already satisfied: petsc in /opt/miniconda3/lib/python3.7/site-packages (3.13.2)
Requirement already satisfied: petsc4py in /opt/miniconda3/lib/python3.7/site-packages (3.13.0)
Requirement already satisfied: numpy in /opt/miniconda3/lib/python3.7/site-packages (from petsc4py) (1.18.1)

**Comment by Marius1311** [comment]
we have a bug on master at the moment - we will fix this and get back to you to look into this a bit more.

**Comment by Marius1311** [comment]
Okay, the bug on master is fixed and we can go ahead. What in particular is it that you want to use cellrank for? From the original scvelo issue, I remember that velocities in the embedding did not coincide with latent time. Are you interested in recovering root and final states in your datasets? Lineage probabilities?

**Comment by ogandril** [comment]
Yes, I would like to get a latent time that matches the velocities in the embedding. Otherwise, I find quite hard to figure out how my cells should be ranked, i.e. by their time or by their position in the velocities fluxes.

**Comment by ogandril** [comment]
Maybe to be more specific: I want to reconstruct a GRN using our WASABI algorithm. This is based upon a time-dependent causality. It works when you have the real time of the process, so I would like to estimate the "real time" when I do not have it. The scVelo paper makes the point that the latent time is closer to real time, so I could use it BUT I am hesitant if it does not match the velocities field.

**Comment by Marius1311** [comment]
Okay, we can try the following: compute the root state(s) of your process using cellrank and feeding this as prior information into latent_time. By explicitly computing the root state, we will gain some understanding of _why_ it is that latent time and your velocity flux disagree.

**Comment by Marius1311** [comment]
To begin, could you please make sure to pull the most recent cellrank version from master, run the code you run to generate the plots you posted above and plot the output of `adata.obs['root_states']`.

**Comment by VolkerBergen** [comment]
Essentially, the issue was the correct identification of root cells which I was hoping could be solved with cellrank. However, the `root_states_probs` look quite mixed as well. @ogandril Have you tried with updated cellrank after bugfixing already? If it still looks incorrect, the root cell index has to be set manually after all. I can then explain how to do so.

**Comment by ogandril** [comment]
I have been updating the latest cellrank. version, rerun the code, and added adata.obs['root_states'].
Result is attached.
![Figure_6](https://user-images.githubusercontent.com/18306816/84245408-47711180-ab05-11ea-9be0-35542b0eba78.png)

**Comment by Marius1311** [comment]
Thanks @ogandril! Looking at your plots, I realised that there is still a bug on master I need to fix. Sorry about this - that's early development ;)

I will let you know once that's done, then you can re-create the plots and we have something to work with.

**Comment by ogandril** [comment]
No problem! Happy to beta test :-)

**Comment by Marius1311** [comment]
Fantastic, great! Our bug on master has just been fixed. The first thing I thought when I saw your plots above was: the identification of the correct number of root states did not work in your case, so I would like to look into that. Can you run the following lines for me please:

```
scv.tl.velocity_graph(adata, n_recurse_neighbors=0, mode_neighbors='connectivities') # recompute the velocity graph
kernel_bwd = cr.tl.transition_matrix(adata, backwards=True) # create a transition matrix based on the velocities for the reverse process
g_bwd = cr.tl.GPCCA(kernel_bwd) # initialise estimator object
g_bwd.compute_schur() # compute partial real, sorted schur decomposition of your trans. mat.
g_bwd.plot_spectrum() # plot the spectrum of your transition matrix
```

Please don't forget to pull from master before doing this and post the plot generated by the last commands

**Comment by ogandril** [comment]
New error: 
>>> kernel_bwd = cr.tl.transition_matrix(adata, backwards=True)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: transition_matrix() got an unexpected keyword argument 'backwards'

(I did pulled from master: Successfully installed cellrank-1.2.1.dev50+gbf2250a

**Comment by Marius1311** [comment]
I misspelled the keyword, it's called `backward` without the `s`. 

I such cases, feel free to consult our API at https://cellrank.readthedocs.io/en/latest/api.html

**Comment by ogandril** [comment]
Here comes the result:
![Figure_1](https://user-images.githubusercontent.com/18306816/84482185-8e8f0c00-ac97-11ea-9304-7b5f35cef0f6.png)

**Comment by Marius1311** [comment]
Thanks! This tells me that there should be only one root state in your data. Using the `g_bwd` object you have initialized above already (re-run the commands above please), please run the following:

```
g_bwd.compute_metastable_states(n_states=1, cluster_key='KEY_TO_YOUR_CLUSTER_LABELS`)
g_bwd.plot_metastable_states(n_cells=30)
```

The `cluster_key` is optional, you can omit it in case you don't have cluster labels. In case you do, these will be used to label to root state, not for computing it.

**Comment by ogandril** [comment]
Here comes the result: 
![Figure_2](https://user-images.githubusercontent.com/18306816/84626918-46b1f400-aee6-11ea-956e-e530b19cf8f5.png)

**Comment by Marius1311** [comment]
Thanks. I suppose this is not what you were expecting?

I think the easiest would be to go over this together in a zoom call, feel free to contact me at marius.lange@helmholtz-muenchen.de

**Comment by ogandril** [comment]
Well I have no real expectation as to what the "correct" roots should be. My main concern is the discrepancy between  latent time and velocity fluxes. 

From our previous discussion, you wrote: "By explicitly computing the root state, we will gain some understanding of why it is that latent time and your velocity flux disagree." 

I would guess the cellranker identification of root state matches mostly the latent time, but maybe, since those are kind of spread out, fluxes do not matches?

PS: I do not use zoom for security reasons. But I can use so many other visio systems :-)

**Comment by ogandril** [comment]
Maybe better explained with this image?

![3289_4](https://user-images.githubusercontent.com/18306816/84662152-b2628400-af1b-11ea-8ea5-849ddce8c49a.png)

**Comment by ogandril** [comment]
If I should draw my expectation, it would look like this:

![expected](https://user-images.githubusercontent.com/18306816/84663350-53057380-af1d-11ea-8ac2-91cfa2371a80.png)

**Comment by Marius1311** [comment]
Okay, so I gather that the computed root state matches your biological expextations mostly, apart from it being a bit spread out?

**Comment by Marius1311** [comment]
and as I said, feel free to email me and we can set up a vc in whatever software you prefer.

**Comment by Marius1311** [comment]
@ogandril , is there anything else we can do for you here? Otherwise, I would go ahead and close this issue.

**Comment by ogandril** [comment]
Well as far as I can tell, the use of cell ranker showed that the latent time starts from the root cells, which is good. But I still do not have a clear view as to why the velocitiy fluxes do NOT start from the rootcells/early latent times....

**Comment by Marius1311** [comment]
There may be several populations of root cells, you can look into this by running 
```
g_bwd.compute_metastable_states(n_states=2, cluster_key='KEY_TO_YOUR_CLUSTER_LABELS`)
g_bwd.plot_metastable_states(n_cells=30)
```
which would do it for 2 populations of root cells.

**Comment by ogandril** [comment]
That's quite interesting. We now identify the "correct" root for the fluxes, (labelled "1"). What is troubling is that now we do not see the previous root cells but a brad new group of cells (labelled "0"). Does that makes sense for you? Naively I would have expected to conserve the previous root and to add a new one.

![2_pop_root_cells](https://user-images.githubusercontent.com/18306816/85281114-bf3a2700-b489-11ea-9f63-4c8c60a79ef2.png)

**Comment by Marius1311** [comment]
Ah, that's interesting and makes a lot of sense to me. Let me try to explain this:

To identify root and final states in your data, we model development as a Markov chain where transition probs are computed on the basis of transcriptomic similarity & RNA velocity. We next seek to coarse grain the state space of this Markov chain, i.e. your set of cells, into a set of 'macrostates' or 'metastable states' of the dynamics. Think of these as groups of cells that, once you enter them, it's unlikely to get out again. Not impossible, but unlikely. Now, this coarse graining isn't just a normal clustering - it's a fuzzy clustering that takes the direction of you Markov chain into account. Fuzzy clustering here means that clusters aren't defined by a hard assignment, but each cell is assigned a probability of being a member of each cluster. What you see above in your plot are the top cells from each fuzzy cluster, ie these are the ones where we are most confident which metastable state they belong to. Note, we infer 'metastable states' - not all of these will be root/final states, we can also identify intermediate metastable states, i.e. checkpoints of development where you stay for a while before moving on. To see how 'stable' your states are, you can plot the coarse grained Markov chain via `g_bwd.plot_coarse_T()`. The states with largest entries on the diagonal will be the most stable ones. 

The above computations rely on a method called [GPCCA](https://www.researchgate.net/publication/325431841_Generalized_Markov_State_Modeling_Method_for_Nonequilibrium_Biomolecular_Dynamics_Exemplified_on_Amyloid_b_Conformational_Dynamics_Driven_by_an_Oscillating_Electric_Field), which we adapted to work for the single cell context, i.e. we made sure it scales to large cell numbers etc. However, that method does **not give you a single  state** - that's by definition not possible for a fuzzy clustering. So, when the user requests one single state, we instead compute the stationary distribution of the Markov chain and just use that. The cells you saw before were just the ones which have highest probability in that stationary distribution. 

So, to move on, I would encourage you to plot the coarse grained transition matrix, so that we can figure out how stable both states are. Also, let's check how dominant they are by plotting
`g_bwd.plot_metastable_states(n_cells=None)`

Once we are happy with the root states, we can compute a latent time that starts from these.

**Comment by ogandril** [comment]
Interesting!!! I am not sure to understand everything but I get the general idea.
So first regarding how 'stable' my states are:
![coarse grained transition matrix](https://user-images.githubusercontent.com/18306816/85418579-3bf40080-b571-11ea-8490-9b0e5054d9a7.png)
And regarding how dominant they are:
![dominance](https://user-images.githubusercontent.com/18306816/85418628-4ca47680-b571-11ea-8b81-a9a17ec7a254.png)

**Comment by Marius1311** [comment]
Nice, thanks for posting this. Keep in mind that we are looking at the backwards process here - the top plot above means that out of the two metastable backward states we have computed, '1' is the more stable one. The lower plot shows you where cells _come from_. It means only few cells are derived from the red '0' root state, most cells come from the blue '1' state. Does this match with your expectations?

**Comment by ogandril** [comment]
Well, once again I do not really have any specific hints as to what the process should be, I'm in an "exploratory mode" here :-)
Do I understand correctly that the blue '1' root state is more stable and generates much more cells that the red '0' root state which seems to be quite localized in the tiny appendix region? If so, could I somehow "force" the velocity fluxes to be computed from the blue '1' root state?

**Comment by ogandril** [comment]
Let me rephrase that: as a matter of fact, that is ALREADY what happened. What we want now is to force the **latent time** to start from the blue '1' root state!

**Comment by Marius1311** [comment]
That sounds much better :) We can't do anything about the velocities because cellrank is **based** on these. However, we can do something about latent time. 

> Well, once again I do not really have any specific hints as to what the process should be, I'm in an "exploratory mode" here :-)
> Do I understand correctly that the blue '1' root state is more stable and generates much more cells that the red '0' root state which seems to be quite localized in the tiny appendix region? 

Yes, that's correct.

**Comment by ogandril** [comment]
I had to change b_bwd to g_bwd for the command to work. And I obtained the attached result, which looks preety good.

![Figure_7](https://user-images.githubusercontent.com/18306816/85712008-28b77100-b6e8-11ea-909b-747f7e84de2e.png)

**Comment by Marius1311** [comment]
Awesome, glad that cellrank worked for you!  Is there any other thing we can help you with?

**Comment by ogandril** [comment]
No, not for the moment, thank's!!! But you were so helpfull that I may chat with you on another occasion :-)
